// @mysten/enoki/dist/cjs/encryption.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var encryption_exports = {};
__export(encryption_exports, {
  createDefaultEncryption: () => createDefaultEncryption,
  createPassthroughEncryption: () => createPassthroughEncryption
});
module.exports = __toCommonJS(encryption_exports);
var import_utils = require("@mysten/sui/utils");
function createDefaultEncryption() {
  async function keyFromPassword(password, salt) {
    const key = await crypto.subtle.importKey(
      "raw",
      new TextEncoder().encode(password),
      { name: "PBKDF2" },
      false,
      ["deriveBits", "deriveKey"]
    );
    const derivedKey = await crypto.subtle.deriveKey(
      {
        name: "PBKDF2",
        salt,
        iterations: 9e5,
        hash: "SHA-256"
      },
      key,
      { name: "AES-GCM", length: 256 },
      false,
      ["encrypt", "decrypt"]
    );
    return { key, derivedKey };
  }
  return {
    async encrypt(password, data) {
      const salt = crypto.getRandomValues(new Uint8Array(16));
      const iv = crypto.getRandomValues(new Uint8Array(12));
      const { derivedKey } = await keyFromPassword(password, salt);
      const payload = await crypto.subtle.encrypt(
        {
          name: "AES-GCM",
          iv
        },
        derivedKey,
        new TextEncoder().encode(data)
      );
      return JSON.stringify({
        payload: (0, import_utils.toBase64)(new Uint8Array(payload)),
        iv: (0, import_utils.toBase64)(iv),
        salt: (0, import_utils.toBase64)(salt)
      });
    },
    async decrypt(password, data) {
      const parsed = JSON.parse(data);
      if (!parsed.payload || !parsed.iv || !parsed.salt) {
        throw new Error("Invalid encrypted data");
      }
      const { derivedKey } = await keyFromPassword(password, (0, import_utils.fromBase64)(parsed.salt));
      const decryptedContent = await crypto.subtle.decrypt(
        {
          name: "AES-GCM",
          iv: (0, import_utils.fromBase64)(parsed.iv)
        },
        derivedKey,
        (0, import_utils.fromBase64)(parsed.payload)
      );
      return new TextDecoder().decode(decryptedContent);
    }
  };
}
function createPassthroughEncryption() {
  return {
    async encrypt(_password, data) {
      return data;
    },
    async decrypt(_password, data) {
      return data;
    }
  };
}
