// asn1-ts/dist/node/macros.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.joint_iso_ccitt = exports.joint_iso_itu_t = exports.iso = exports.itu_r = exports.ccitt = exports.itu_t = exports.NOT_A_NUMBER = exports.MINUS_INFINITY = exports.PLUS_INFINITY = exports.FALSE_BIT = exports.TRUE_BIT = exports.FALSE = exports.TRUE = exports.CharacterString = void 0;
var CharacterString_1 = require("./types/CharacterString");
Object.defineProperty(exports, "CharacterString", { enumerable: true, get: function () { return __importDefault(CharacterString_1).default; } });
exports.TRUE = true;
exports.FALSE = false;
exports.TRUE_BIT = 1;
exports.FALSE_BIT = 0;
exports.PLUS_INFINITY = Infinity;
exports.MINUS_INFINITY = -Infinity;
exports.NOT_A_NUMBER = NaN;
exports.itu_t = 0;
exports.ccitt = 0;
exports.itu_r = 0;
exports.iso = 1;
exports.joint_iso_itu_t = 2;
exports.joint_iso_ccitt = 2;
