// @mysten/signers/dist/cjs/ledger/bcs.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var bcs_exports = {};
__export(bcs_exports, {
  SuiMoveObject: () => SuiMoveObject
});
module.exports = __toCommonJS(bcs_exports);
var import_bcs = require("@mysten/sui/bcs");
var import_utils = require("@mysten/sui/utils");
const SUI_FRAMEWORK_ADDRESS = (0, import_utils.normalizeSuiAddress)("0x2");
const SUI_SYSTEM_ADDRESS = (0, import_utils.normalizeSuiAddress)("0x3");
const MoveObjectType = import_bcs.bcs.enum("MoveObjectType", {
  Other: import_bcs.bcs.StructTag,
  GasCoin: null,
  StakedSui: null,
  Coin: import_bcs.bcs.TypeTag
});
const SuiMoveObject = import_bcs.bcs.struct("SuiMoveObject", {
  data: import_bcs.bcs.enum("Data", {
    MoveObject: import_bcs.bcs.struct("MoveObject", {
      type: MoveObjectType.transform({
        input: (objectType) => {
          const structTag = (0, import_utils.parseStructTag)(objectType);
          if (structTag.address === SUI_FRAMEWORK_ADDRESS && structTag.module === "coin" && structTag.name === "Coin" && typeof structTag.typeParams[0] === "object") {
            const innerStructTag = structTag.typeParams[0];
            if (innerStructTag.address === SUI_FRAMEWORK_ADDRESS && innerStructTag.module === "sui" && innerStructTag.name === "SUI") {
              return { GasCoin: true, $kind: "GasCoin" };
            }
            return { Coin: (0, import_utils.normalizeStructTag)(innerStructTag), $kind: "Coin" };
          } else if (structTag.address === SUI_SYSTEM_ADDRESS && structTag.module === "staking_pool" && structTag.name === "StakedSui") {
            return { StakedSui: true, $kind: "StakedSui" };
          }
          return {
            Other: {
              ...structTag,
              typeParams: structTag.typeParams.map((typeParam) => {
                return import_bcs.TypeTagSerializer.parseFromStr((0, import_utils.normalizeStructTag)(typeParam));
              })
            },
            $kind: "Other"
          };
        }
      }),
      hasPublicTransfer: import_bcs.bcs.bool(),
      version: import_bcs.bcs.u64(),
      contents: import_bcs.bcs.byteVector().transform({ input: import_utils.fromBase64 })
    })
  }),
  owner: import_bcs.bcs.Owner.transform({
    input: (objectOwner) => {
      if (objectOwner === "Immutable") {
        return { Immutable: null };
      } else if ("Shared" in objectOwner) {
        return { Shared: { initialSharedVersion: objectOwner.Shared.initial_shared_version } };
      } else if ("ConsensusAddressOwner" in objectOwner) {
        return {
          ConsensusAddressOwner: {
            owner: objectOwner.ConsensusAddressOwner.owner,
            startVersion: objectOwner.ConsensusAddressOwner.start_version
          }
        };
      }
      return objectOwner;
    }
  }),
  previousTransaction: import_bcs.bcs.ObjectDigest,
  storageRebate: import_bcs.bcs.u64()
});
